### PR TITLE
support mtp

### PR DESCRIPTION
**Pull request by liupeng374**
## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Review by xiaobaicxy — APPROVED** [review]

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 1]
```diff
@@ -92,14 +93,21 @@ def __init__(self, model_runner: ModelRunner):
         self.max_context_len = model_runner.model_config.context_len
         self.req_to_token = model_runner.req_to_token_pool.req_to_token
         self.graph_mode = False
+        self.speculative_num_draft_tokens = (
+            model_runner.server_args.speculative_num_draft_tokens
+        )
+        self.mtp_mask = torch.tril(torch.ones(2048, 2048, dtype=torch.bool)).npu()
```
Comment (xiaobaicxy) [inline python/sglang/srt/layers/attention/ascend_backend.py:99]: Make mask a class variable so that memory is allocated only once

File: python/sglang/srt/speculative/build_eagle_tree.py [hunk 2]
```diff
@@ -1,18 +1,137 @@
...
+    topk: int,
+    draft_token_num: int,
+    tree_mask_mode: int,
+    bs: int,
+):
+    # Generate batch and token index ranges
+    bs_range = torch.arange(bs, device=parent_list.device).view(-1, 1)
+    draft_token_num_range = torch.arange(draft_token_num, device=parent_list.device)
+
+    if draft_token_num == 2 and topk == 1 and tree_mask_mode == TreeMaskMode.FULL_MASK:
```
Comment (chenxu140) [inline python/sglang/srt/speculative/build_eagle_tree.py:37]: It is recommended to use comments to explain the meaning of this special processing section.

File: python/sglang/srt/models/deepseek_nextn.py [hunk 3]
```diff
@@ -70,7 +70,7 @@ def __init__(
             0,
             quant_config=quant_config,
             is_nextn=True,
-            prefix=add_prefix("decoder", prefix),
+            prefix=add_prefix("layers.61", prefix),
```
Comment (xiaobaicxy) [inline python/sglang/srt/models/deepseek_nextn.py:73]: It will cause NV and others to fail.

File: python/sglang/srt/layers/attention/ascend_backend.py (hunk 1 above)
Comment (liupeng374) [reply python/sglang/srt/layers/attention/ascend_backend.py:99]: attention backend only need init once, so the tensor is allocated only once

File: python/sglang/srt/managers/schedule_batch.py [hunk 4]
```diff
@@ -885,7 +885,7 @@ class ScheduleBatch(ScheduleBatchDisaggregationDecodeMixin):
     has_grammar: bool = False
 
     # Device
-    device: str = "cuda"
+    device: str = "npu"
```
[the remark this answers is not in the document]
Comment (liupeng374) [reply python/sglang/srt/managers/schedule_batch.py]: ok, already fix

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 5]
```diff
@@ -233,6 +254,133 @@ def forward_extend(
             attn_output = attn_output.reshape(-1, layer.tp_q_head_num, layer.v_head_dim)
             return attn_output
 
+    def forward_mtp(self, q, k, v, layer: RadixAttention, forward_batch: ForwardBatch):
+        q, q_rope = q
+        if not self.graph_mode:
+            num_token_padding = q.shape[0]
+            q = q[: forward_batch.num_token_non_padded_cpu]
+            q_rope = q_rope[: forward_batch.num_token_non_padded_cpu]
+        c_kv, k_rope = forward_batch.token_to_kv_pool.get_kv_buffer(layer.layer_id)
```
[the remark this answers is not in the document]
Comment (liupeng374) [reply python/sglang/srt/layers/attention/ascend_backend.py:292]: ok, I rebased the newest codes

File: python/sglang/srt/layers/attention/ascend_backend.py [hunk 6]
```diff
@@ -233,6 +247,69 @@ def forward_extend(
...
+        if self.forward_metadata.seq_lens_cpu_int is None:
+            actual_seq_lengths_kv = self.forward_metadata.seq_lens_cpu_list
+        else:
+            actual_seq_lengths_kv = (
+                self.forward_metadata.seq_lens_cpu_int.cpu().int().tolist()
+            )
+        q_nope = q_nope.view(-1, layer.tp_q_head_num, q_nope.shape[-1])
+        q_rope = q_rope.view(-1, layer.tp_q_head_num, q_rope.shape[-1])
+        if forward_batch.forward_mode.is_target_verify():
+            actual_seq_lengths = np.arange(2, 2 + q_nope.shape[0], 2)
```
[the remark this answers is not in the document]
Comment (liupeng374) [reply python/sglang/srt/layers/attention/ascend_backend.py]: ok, this is speculative_num_draft_tokens

File: python/sglang/srt/speculative/eagle_draft_extend_npu_graph_runner.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+"""Run the model with npu graph and torch.compile"""
+
+from __future__ import annotations
+
+import threading
+from typing import TYPE_CHECKING
+
+import torch
+
+from python.sglang.srt.speculative.eagle_draft_extend_cuda_graph_runner import (
```
[the remark this answers is not in the document]
Comment (liupeng374) [reply python/sglang/srt/speculative/eagle_draft_extend_npu_graph_runner.py]: done

File: python/sglang/srt/speculative/eagle_worker.py [hunk 8]
```diff
@@ -7,6 +7,12 @@
 import torch
 from huggingface_hub import snapshot_download
 
+from python.sglang.srt.speculative.eagle_draft_extend_npu_graph_runner import (
```
[the remark this answers is not in the document]
Comment (liupeng374) [reply python/sglang/srt/speculative/eagle_worker.py]: done

File: python/sglang/srt/speculative/build_eagle_tree.py (hunk 2 above)
Comment (liupeng374) [reply python/sglang/srt/speculative/build_eagle_tree.py:38]: done

**Review by chenxu140 — APPROVED** [review]

File: python/sglang/srt/layers/quantization/w8a8_int8.py [hunk 9]
```diff
@@ -256,6 +256,8 @@ def get_quant_method(
 
         if _is_npu:
             if isinstance(layer, LinearBase):
+                if "decoder" in prefix:
+                    prefix = prefix.replace("decoder", "layers.61")
```
Comment (xiaobaicxy) [inline python/sglang/srt/layers/quantization/w8a8_int8.py:260]: It cannot support multi-mtp layers, or other modes that include mtp, eg a model has 80 layers, and mtp is the 81th layer

**Review by xiaobaicxy — APPROVED** [review]

File: python/sglang/srt/layers/attention/ascend_backend.py (hunk 1 above)
Comment (ranjiewen) [reply python/sglang/srt/layers/attention/ascend_backend.py:99]: use mask = self.attn_mask_builder.get_attn_mask(2048, q.dtype, q.device).to(
                torch.int8
            )

**Review by ranjiewen — APPROVED** [review]
